Source/Core/VideoCommon/TextureInfo.cpp

```diff
@@ -0,0 +1,325 @@
+// Copyright 2021 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "VideoCommon/TextureInfo.h"
+
+#include <xxhash.h>
+
+#include "Common/Align.h"
+#include "Core/HW/Memmap.h"
+#include "VideoCommon/BPMemory.h"
+#include "VideoCommon/SamplerCommon.h"
+#include "VideoCommon/TextureDecoder.h"
+
+TextureInfo::TextureInfo(u32 stage)
+{
+  const FourTexUnits& tex = bpmem.tex[stage >> 2];
+  const u32 id = stage & 3;
+
+  m_texture_format = tex.texImage0[id].format;
+  m_tlut_format = tex.texTlut[id].tlut_format;
+
+  const bool is_palette_texture = IsColorIndexed(m_texture_format);
+  if (is_palette_texture)
+    m_palette_size = TexDecoder_GetPaletteSize(m_texture_format);
+
+  m_raw_width = tex.texImage0[id].width + 1;
+  m_raw_height = tex.texImage0[id].height + 1;
+
+  // TexelSizeInNibbles(format) * width * height / 16;
+  m_block_width = TexDecoder_GetBlockWidthInTexels(m_texture_format);
+  m_block_height = TexDecoder_GetBlockHeightInTexels(m_texture_format);
+
+  m_expanded_width = Common::AlignUp(m_raw_width, m_block_width);
+  m_expanded_height = Common::AlignUp(m_raw_height, m_block_height);
+
+  m_texture_size =
+      TexDecoder_GetTextureSizeInBytes(m_expanded_width, m_expanded_height, m_texture_format);
+
+  m_raw_address = (tex.texImage3[id].image_base /* & 0x1FFFFF*/) << 5;
+
+  m_from_tmem = tex.texImage1[id].cache_manually_managed != 0;
+  const u32 tmem_address_even = m_from_tmem ? tex.texImage1[id].tmem_even * TMEM_LINE_SIZE : 0;
+  const u32 tmem_address_odd = m_from_tmem ? tex.texImage2[id].tmem_odd * TMEM_LINE_SIZE : 0;
+  if (m_from_tmem)
+  {
+    m_data = &texMem[tmem_address_even];
+    m_tmem_odd = &texMem[tmem_address_odd];
+  }
+  else
+  {
+    m_data = Memory::GetPointer(m_raw_address);
```

Comment:
`m_data` is a bad name. Implies that it's a owned copy of the data, while this is just a pointer. The underlying data might change.

`m_ptr` might be better? 

Or a `u32 m_address` and every usage can just do a `Memory::GetPointer(m_address)`?

If *somebody* ever gets around to implementing locking (*glares at self*), then storing pointers into emulated memory is a bad idea that can potentially break things.

Reply:
Yeah, I think I'll stick with `m_ptr` for now.  Only because certain things become a bit more painful once we say we can't store pointers.  If locking manifests, we will definitely want to revisit.